tidy3d/plugins/waveguide/rectangular_dielectric.py

```diff
@@ -0,0 +1,944 @@
+        result = numpy.array(val, ndmin=1)
+        if any(result < 0):
+            raise ValidationError("Values may not be negative.")
+        return result
+
+    @pydantic.validator("box_medium", always=True)
+    def _set_box_medium(cls, val, values):
+        return values["clad_medium"] if val is None else val
+
+    @pydantic.validator("clad_thickness", always=True)
+    def _set_clad_thickness(cls, val, values):
+        if val is None:
+            wavelength = values["wavelength"]
+            medium = values["clad_medium"]
+            n = numpy.array([medium.nk_model(f)[0] for f in C_0 / wavelength])
+            lda = wavelength / n
+            return 1.5 * lda.max()
+        return val
+
+    @pydantic.validator("box_thickness", always=True)
+    def _set_box_thickness(cls, val, values):
+        if val is None:
+            wavelength = values["wavelength"]
+            medium = values["box_medium"]
+            n = numpy.array([medium.nk_model(f)[0] for f in C_0 / wavelength])
+            lda = wavelength / n
+            return 1.5 * lda.max()
+        return val
+
+    @pydantic.validator("side_margin", always=True)
+    def _set_side_thickness(cls, val, values):
+        return max(values["clad_thickness"], values["box_thickness"]) if val is None else val
+
+    @pydantic.validator("gap", always=True)
+    def _validate_gaps(cls, val, values):
+        if val.size == 1 and values["core_width"].size != 2:
+            return numpy.array([val[0]] * (values["core_width"].size - 1))
+        if val.size != values["core_width"].size - 1:
+            raise ValidationError("Number of gaps must be 1 less than number of core widths.")
+        return val
+
+    @pydantic.root_validator
+    def _ensure_consistency(cls, values):
+        if values["sidewall_thickness"] > 0 and values["sidewall_medium"] is None:
+            raise ValidationError(
+                "Sidewall medium must be provided when sidewall thickness is greater than 0."
+            )
+
+        if values["sidewall_thickness"] == 0 and values["sidewall_medium"] is not None:
+            log.warning("Sidewall medium not used because sidewall thickness is zero.")
+
+        if values["surface_thickness"] > 0 and values["surface_medium"] is None:
+            raise ValidationError(
+                "Surface medium must be provided when surface thickness is greater than 0."
+            )
+
+        if values["surface_thickness"] == 0 and values["surface_medium"] is not None:
+            log.warning("Surface medium not used because surface thickness is zero.")
+
+        if values["propagation_axis"] == values["normal_axis"]:
+            raise ValidationError("Propagation and normal axes must be different.")
+
+        return values
+
+    @cached_property
+    def lateral_axis(self):
+        """Lateral direction axis"""
+        return 3 - self.propagation_axis - self.normal_axis
+
+    def _swizzle(self, lateral_coord, normal_coord, propagation_coord):
```

Comment:
interesting method name 😆 . I like it, but would maybe suggest that we use something more boring but also more descriptive. `_set_waveguide_coords`? Also, add type hints please to this and the other ones. and add a periodic at the end of docstrings.

Reply:
I know the name [from OpenGL](https://www.khronos.org/opengl/wiki/Data_Type_(GLSL)#Swizzling), it's not my creation. That's a known operation to swap the axis order or add/remove/duplicate components. So, for the sake of clarity, I renamed to `_swap_axis`